### PR TITLE
Switching pattern categories inserter to Tabs component with arrow key navigation

### LINKED ISSUE #60209: Improve Keyboard Navigation on Side Panel Inserter
> the content of the three tab panels uses different design and interaction: IMHO this doesn't help a predictable, consistent, experience for users
> @afercia https://github.com/WordPress/gutenberg/issues/24975

The keyboard navigation between the side panel inserter has inconsistent interactions and makes it hard to use. With a little refinement, I think it would be quite nice to interact with. The panels _mostly_ use a pattern of: Tab to area, then use arrows within the area. 

Patterns inserter view

However, the Patterns and Media panels use Tab to navigate between lists of buttons to drill down to further layers. On one hand it makes sense since Tab should generally navigate between the buttons. In practice, I find it frustrating because l try to use the arrows to navigate this area like all the others do. This Tab to navigate within the section issue becomes exacerbated when you have a selected Pattern. 

My expected behavior for the Pattern panel:
- Select Pattern Tab
- Focus moves into Patterns panel
- Use arrows to navigate between the patterns
- Select a pattern
- Focus moves into the Patterns for that category
- Arrow between patterns
- Shift+Tab to return to my selected pattern category
- Arrow to a new pattern category, repeat

How it works now:
- Select Pattern Tab
- Focus moves into Patterns panel
- Use Tab to navigate between the patterns
- Select a pattern
- Focus moves into the Patterns for that category
- Arrow between patterns
- Shift+Tab to  try to return to my selected pattern category
- Instead, I'm at the _bottom_ of the Patterns category list
- Shift + Tab 15x times to return to the Tabs

Other issues with the current behavior:
- Because focused is moved into the individual patterns sidebar when it is selected/revealed, you can't move focus into a selected pattern by pressing the enter key on that category.

https://github.com/WordPress/gutenberg/assets/967608/ce7ef301-1b79-4aa6-8e7b-7f8be603be32

## What is your proposed solution?

Adopt a consistent pattern of using Tab to navigate between sections and arrows within the sections to make the sidebar block/pattern inserter more usable.

### PR DESCRIPTION
**Pull request by jeryj**
Fixes https://github.com/WordPress/gutenberg/issues/60209

## What?

Improves the accessibility and keyboard navigation of using the pattern categories list in the inserter.

## Why?

The current implementation is cumbersome with too many tabstops and the panel is disconnected from the activating button, making it hard to switch between categories using a keyboard.

## How?

Changes the structure of the pattern categories from a listbox of buttons that open a panel to Tabs implementation that considers all the categories as one tabstop with arrow keys to move between the categories.

### Testing Instructions for Keyboard

- Open Block inserter (in post editor and in site editor)
- Select Pattern Tab
- Tab to patten categories panel
- Use arrows to navigate between the patterns
- Select a pattern
- Tab into patterns panel
- Shift + Tab out
- Focus should be on the selected category
- Arrow between patterns and repeat switching between categories

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/967608/e06aa513-cd78-4722-8172-239fa126e849

### CODE REVIEW

**Comment by jeryj** [comment]
There's a bug somewhere where:
- if a block is selected
- a pattern category is clicked
- the inserter closes and focus goes to the block

**Comment by draganescu** [comment]
I added the needs design feedback as I am not sure if opening the 1st category directly is a good interaction. I fear people won't bother checking out the other categories and will assume that what has opened _are all of the patterns_.

**Review by draganescu — COMMENTED** [review]
I tested this and I think it's a big improvement:

- the focus remains on the selected category
- the up and down arrows move selection to a new category
- the list of categories does not scroll on chaning categories

The current implementation is quite unpredictable and even with the mouse the focus capture by the category filter button is disorientating.

**Comment by jameskoster** [comment]
I'm also not convinced about auto-opening the first category. Probably one to extract and explore in detail separately?

**Comment by jeryj** [comment]
Regarding the pattern category open or closed automatically...

Either I'm doing it wrong, or there is a bug where you can't have the first item of the tab panel category focused _without also being active_. I've [asked the components team about this behavior](https://github.com/WordPress/gutenberg/issues/52997#issuecomment-2027315542).

> I fear people won't bother checking out the other categories and will assume that what has opened are all of the patterns.

Isn't the All tab all of patterns in a paginated list?

**Comment by scruffian** [comment]
> On this PR, the first pattern category is open when the top level pattern tab is selected.

This seems to no longer be the case in this PR

File: packages/block-editor/README.md [hunk 1]
```diff
@@ -1032,6 +1032,10 @@ _Returns_
 
 A hook used to set the editor mode to zoomed out mode, invoking the hook sets the mode.
 
+_Parameters_
+
+-   _zoomOut_ `boolean`: If we should enter into zoomOut mode or not
```
Comment (scruffian) [inline packages/block-editor/README.md:1037]: This doesn't seem to be used anywhere.

File: packages/block-editor/src/components/inserter/menu.js [hunk 2]
```diff
@@ -215,6 +191,9 @@ function InserterMenu(
 		! delayedFilterValue &&
 		selectedMediaCategory;
 
+	// When the pattern panel is showing, we want to use zoom out mode
+	useZoomOut( showPatternPanel );
```
Comment (scruffian) [inline packages/block-editor/src/components/inserter/menu.js:233]: Why have we moved this here, rather than keeping it inside `PatternCategoryPreviewPanel`?

Comment (jeryj) [reply packages/block-editor/src/components/inserter/menu.js:233]: Before, the `PatternCategoryPreviewPanel` was mounted and then the patterns switched out inside of it. Now, each panel is its own component (via the tabs component). So, if we left it within `PatternCategoryPreviewPanel` it would rerun the hook each time you switched categories, toggling between zoom out and edit. By moving it up a level we can preserve the right mode.

**Comment by jeryj** [comment]
> This seems to no longer be the case in this PR

Yup - changed it yesterday since that's the behavior people think will be better. Forgot to update the PR description. Current status is waiting on response from the components team about how to move focus to the first selected tab but _not_ activate it.

**Comment by draganescu** [comment]
I removed the design feedback label because it seems we don't select the 1st tab anymore. Now focus is on categories container.

**Comment by draganescu** [comment]
@jeryj can we just skip the tab to the category container, and make the flow:

- tab to tabs
- arrow key right to patterns
- enter **focus remains on the patterns tab**
- tab to 1st category
- enter

**Comment by jeryj** [comment]
> can we just skip the tab to the category container, and make the flow:
> - tab to tabs
> - arrow key right to patterns
> - enter focus remains on the patterns tab
> - tab to 1st category
> - enter

Yes, that's what I'd like to happen as well. The current set-up of the tabs v2 component moves focus to the tabs container and not the 1st category though. That step is what I want to change so focus is moved to the first category instead of the tabs container. I [asked the components team here and explained this in more depth](https://github.com/WordPress/gutenberg/issues/52997#issuecomment-2035568332).

**Comment by alexstine** [comment]
I left a comment on the linked issue. Ultimately, there needs to be a default tab selected or the current state of the code forces the wrapper to have focus. Keep in mind, when a default tab is not chosen, all tabs get `tabindex="-1"` which means even if you did set `focusable` to false to remove `tabindex="0"` from the wrapper, you would create a situation where all tabs are now inaccessible as `tabindex="-1"` removes from tab order.

**Comment by jeryj** [comment]
The behavior of the patterns sidebar now matches trunk. You have to press enter on your selected tab to see the patterns for that category.

File: packages/block-editor/src/components/inserter/block-patterns-tab/index.js [hunk 3]
```diff
@@ -74,39 +80,47 @@ function BlockPatternsTab( {
...
+								showTitlesAsTooltip={ false }
+							/>
+						</div>
 					) }
 				</MobileTabNavigation>
 			) }
 			{ showPatternsExplorer && (
 				<PatternsExplorerModal
-					initialCategory={ initialCategory }
+					initialCategory={ selectedCategory || categories[ 0 ] }
```
Comment (getdave) [inline packages/block-editor/src/components/inserter/block-patterns-tab/index.js:123]: Are `categories` guaranteed to exist / be an array?

File: packages/block-editor/src/components/inserter/block-patterns-tab/pattern-category-preview-panel.js [hunk 4]
```diff
@@ -20,34 +12,15 @@ export function PatternCategoryPreviewPanel( {
...
-			const [ firstTabbable ] = focus.tabbable.find( container.current );
-			firstTabbable?.focus();
-		} );
-		return () => clearTimeout( timeout );
-	}, [ category ] );
-
-	// Move to zoom out mode when this component is mounted
-	// and back to the previous mode when unmounted.
-	useZoomOut();
-
```
Comment (getdave) [inline packages/block-editor/src/components/inserter/block-patterns-tab/pattern-category-preview-panel.js:36]: Can you explain why it was ok to remove all this code? It seems it was handling zoom out and autofocus.

File: packages/block-editor/src/components/inserter/menu.js [hunk 5]
```diff
@@ -223,8 +202,50 @@ function InserterMenu(
 		setSelectedTab( value );
 	};
 
+	const patternsTab = useMemo(
```
Comment (getdave) [inline packages/block-editor/src/components/inserter/menu.js]: Can you explain the choice to `useMemo` here?

File: packages/block-editor/src/hooks/use-zoom-out.js [hunk 6]
```diff
@@ -11,26 +11,36 @@ import { store as blockEditorStore } from '../store';
...
-			return () => {
-				// Revert to original mode
-				__unstableSetEditorMode( mode );
-			};
+		} else if ( ! zoomOut && originalEditingMode.current !== mode ) {
+			__unstableSetEditorMode( originalEditingMode.current );
 		}
-		// eslint-disable-next-line react-hooks/exhaustive-deps
-	}, [] );
+	}, [ zoomOut, mode ] );
```
Comment (getdave) [inline packages/block-editor/src/hooks/use-zoom-out.js]:
```suggestion
	}, [ __unstableSetEditorMode, zoomOut, mode ] );
```

File: packages/block-editor/src/hooks/use-zoom-out.js [hunk 7]
```diff
@@ -11,26 +11,36 @@ import { store as blockEditorStore } from '../store';
...
+			originalEditingMode.current = mode;
+		}
+
+		return () => {
+			// We need to use  __unstableGetEditorMode() here and not `mode`, as mode may not update on unmount
+			if ( __unstableGetEditorMode() !== originalEditingMode.current ) {
+				__unstableSetEditorMode( originalEditingMode.current );
+			}
 		};
 	}, [] );
```
Comment (getdave) [inline packages/block-editor/src/hooks/use-zoom-out.js:36]:
```suggestion
	}, [ __unstableGetEditorMode, __unstableSetEditorMode, mode ] );
```

File: packages/block-editor/README.md [hunk 8]
```diff
@@ -1032,6 +1032,10 @@ _Returns_
...
+
```
Comment (getdave) [inline packages/block-editor/README.md:1038]: Can you help me understand why this change is part of this PR? What required the addition of this new param?

**Review by getdave — COMMENTED** [review]
Thanks for working on consistency in the editor 👍 

There are a lot of changes in this PR which meant that I found it quite hard to review. That said, I recognise that this is completely transplanting one UI component for another and so I guess there was a need for a fair amount of small changes across a lot of files to facilitate this. If there is a way to break this down it might be better to land in smaller chunks but I appreciate that might not be practical.

Some things I noticed:

- There doesn't appear to be any visual focus on the pattern items. Did we loose this?
<details>
<summary>Video: no visual focus</summary>

https://github.com/WordPress/gutenberg/assets/444434/b5187bc7-8bb1-496b-8af2-0923577f35ba

</details>

- Unsure if it's related but the animation to enter zoom out mode when select a pattern category is jarring. If it's not a bug in this PR it needs to have an Issue to track it because it's noticeable.
<details>
<summary>Video: janky animation</summary>

https://github.com/WordPress/gutenberg/assets/444434/d2b2d977-3b2e-4cdc-a501-1aeb20c2ce50

</details>

- I expected to be able to use the right arrow key to move into the pattern category. Instead I had to hit enter. I'm not sure if it's the correct paradigm but it felt natural to me that if I'm using arrows to move up and down between categories I should also be able to use the arrow to enter the category itself.
- Once I select a pattern category with `ENTER` I need to hit `TAB` to move into the panel. Is that right? It felt as though focus should be auto-transferred. I noticed [this comment](https://github.com/WordPress/gutenberg/pull/60257#issuecomment-2059869323) so I assume that's expected behaviour.


Overall this is better than what we have on `trunk` right now. If we can sort the visual focus for keyboard I'd be ok with merging and iterating.

File: packages/block-editor/src/components/inserter/block-patterns-tab/index.js (hunk 3 above)
Comment (jeryj) [reply packages/block-editor/src/components/inserter/block-patterns-tab/index.js:123]: I believe so. I didn't change how `categories` is used. I [matched what `trunk` did](https://github.com/WordPress/gutenberg/pull/60257/files#diff-d2d1d0f46309ac48c10aa1db0165133f00cbcaa415f9772743fc7d029a7bb345L34).

File: packages/block-editor/src/components/inserter/menu.js (hunk 5 above)
Comment (jeryj) [reply packages/block-editor/src/components/inserter/menu.js]: It was memoized before: https://github.com/WordPress/gutenberg/pull/60257/files#diff-2745181e57f35c49a91ec773fd2b1740c75f87b0f74efb857dd7a08a90b064e7L158.

I can't remember why I moved it down in the file though. I'll see if I can move it closer to its previous position so that relationship in git is clearer.

File: packages/block-editor/src/components/inserter/block-patterns-tab/pattern-category-preview-panel.js (hunk 4 above)
Comment (jeryj) [reply packages/block-editor/src/components/inserter/block-patterns-tab/pattern-category-preview-panel.js:36]: This did handle autofocus and zoom out. I'd like to remove autofocus, as none of the other tab components move focus into the panel, nor is it mentioned on [the tabs pattern ARIA guide](https://www.w3.org/WAI/ARIA/apg/patterns/tabs/). I played around with it a bit on the keyboard and quite liked not moving autofocus for being able to activate and see the patterns in a category without committing to the category. It makes browsing categories a lot easier IMO.

The zoom out mode hook had to be moved up a level to the menu.js.

File: packages/block-editor/README.md (hunk 8 above)
Comment (jeryj) [reply packages/block-editor/README.md:1038]: Previously the `useZoomOut` hook toggled between zoom in/out mode each time it ran. Passing a param of true/false allows us to intentionally choose if we want to use zoom out mode or not. 

This param wasn't necessary in the previous set-up because the pattern category panel would unmount (turning zoom out mode off) then re mount a new category (turning zoom out mode back on). There's [more context in this comment.](https://github.com/WordPress/gutenberg/pull/60257/files#r1553840557)

**Comment by jeryj** [comment]
> There doesn't appear to be any visual focus on the pattern items. Did we loose this?

Nope, it's missing on trunk too. Created an issue: https://github.com/WordPress/gutenberg/issues/60872

> Unsure if it's related but the animation to enter zoom out mode when select a pattern category is jarring. If it's not a bug in this PR it needs to have an Issue to track it because it's noticeable.

Also the same animation on `trunk`.

> I expected to be able to use the right arrow key to move into the pattern category. Instead I had to hit enter. I'm not sure if it's the correct paradigm but it felt natural to me that if I'm using arrows to move up and down between categories I should also be able to use the arrow to enter the category itself.

I can see what you mean, but using focus to move into tabs isn't a standard pattern. There was talk about making this into a menu instead in the future. Using the arrows in that way would align more with menu pattern semantics. The tab semantics don't expect an arrow into the panel interaction though. 

> Once I select a pattern category with ENTER I need to hit TAB to move into the panel. Is that right? It felt as though focus should be auto-transferred. I noticed https://github.com/WordPress/gutenberg/pull/60257#issuecomment-2059869323 so I assume that's expected behaviour.

Explained my take on removing autofocus here: https://github.com/WordPress/gutenberg/pull/60257#discussion_r1570953362

> Overall this is better than what we have on trunk right now. If we can sort the visual focus for keyboard I'd be ok with merging and iterating.

Let's merge, as visual focus is an issue on trunk as well. The issue is created (https://github.com/WordPress/gutenberg/issues/60872) and I'd rather not expand this PR.

File: packages/block-editor/src/components/inserter/menu.js [hunk 9]
```diff
@@ -224,7 +241,11 @@ function InserterMenu(
 	};
 
 	return (
-		<div className="block-editor-inserter__menu">
+		<div
+			className={ classnames( 'block-editor-inserter__menu', {
+				'show-panel': showPatternPanel,
```
Comment (draganescu) [inline packages/block-editor/src/components/inserter/menu.js:246]: I don't understand why this class name is `show-panel` and the toggle bool is `showPatternPanel` - shouldn't the class be about pattern panel as well?

**Review by draganescu — APPROVED** [review]
When switching to a category the whole sidebar is animated again this time with the preview open. It seems unrelated to this PR?

Did we get a tentative 🆗 from @WordPress/gutenberg-components team?

File: packages/block-editor/src/components/inserter/menu.js (hunk 9 above)
Comment (jeryj) [reply packages/block-editor/src/components/inserter/menu.js:246]: I left it generic, as the panel is a fly out of the menu for anything to get put into. For example, the media tab should get changed to this pattern as well. We could also add a secondary class that says it's a pattern panel if needed.

**Comment by jeryj** [comment]
> Did we get a tentative 🆗 from @WordPress/gutenberg-components team?

Yes, the new active id fallback behavior was approved in https://github.com/WordPress/gutenberg/pull/60681 and approval for using the tabs here via https://github.com/WordPress/gutenberg/pull/60573#pullrequestreview-1992347666

**Comment by jeryj** [comment]
> When switching to a category the whole sidebar is animated again this time with the preview open. It seems unrelated to this PR?

I see what you mean. It's on Trunk too.

**Comment by youknowriad** [comment]
Did you consider looking at the "Media" tab as well which I believe was modeled to match the patterns inserter. It would be weird to have two different ways of implementing the same UI.

**Comment by youknowriad** [comment]
Also @ntsekouras might be interested in this PR as well.

**Comment by jeryj** [comment]
> Did you consider looking at the "Media" tab as well which I believe was modeled to match the patterns inserter. It would be weird to have two different ways of implementing the same UI.

@youknowriad - Yes, I think they should share the same interactions.

**Comment by ntsekouras** [comment]
>@youknowriad - Yes, I think they should share the same interactions.

That sounds great! I'm happy to help with reviews to apply it in media tab for consistency. It should be a very similar changeset.
